pdns/histogram.hh

```diff
@@ -0,0 +1,137 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include <algorithm>
+#include <atomic>
+#include <vector>
+
+namespace pdns {
+
+// By convention, we are using microsecond units
+struct Bucket
+{
+  std::string d_name;
+  uint64_t d_boundary{0};
+  uint64_t d_count{0};
+};
+
+inline bool operator<(uint64_t b, const Bucket& bu)
+{
+  // we are using less-or-equal
+  return b <= bu.d_boundary;
```

Comment:
I find that a bit confusing. I wonder if it would not be clearer to define a different comparison operator (`lessOrEqual`?) and pass that to `std::upper_bound` and `std::is_sorted`?

Reply:
Not needed for `std::is_sorted`, it's comparing `uints64_t`. Actually it should also check if each bound is unique.